File: claude_standup/reporter.py
"""Reporter module — generates standup reports from classified activities.

Two modes:
- Template: instant local report (generate_template_report)
- LLM: one `claude -p` call to polish the report (generate_llm_report)
"""

from __future__ import annotations

import itertools
import operator
import shutil
import subprocess

from claude_standup.models import Activity


# ---------------------------------------------------------------------------
# Template reporter — local, instant report (no LLM)
# ---------------------------------------------------------------------------

def _format_activity_line(act: Activity) -> str:
    """Format a single activity into a human-readable bullet line."""
    org_repo = ""
    if act.git_org and act.git_repo:
        org_repo = f"({act.git_org}/{act.git_repo})"
    elif act.git_org:
        org_repo = f"({act.git_org})"
    time_part = f" ~{act.time_spent_minutes}min" if act.time_spent_minutes else ""
    return f"[{act.classification}]{org_repo} {act.summary}{time_part}"


def _template_markdown(
    activities: list[Activity],
    pending_count: int,
    lang: str,
) -> str:
    """Render activities as a Markdown template report."""
    sections: list[str] = []

    sorted_acts = sorted(activities, key=operator.attrgetter("day"))
    for day, group in itertools.groupby(sorted_acts, key=operator.attrgetter("day")):
        lines: list[str] = [f"## {day}", ""]
        lines.append("### Done" if lang == "en" else "### Done")
        for act in group:
            lines.append(f"- {_format_activity_line(act)}")
        sections.append("\n".join(lines))

    body = "\n\n".join(sections)

    if pending_count > 0:
        if lang == "en":
            pending_section = (
                f"### Pending classification\n"
                f"- {pending_count} sessions pending classification"
            )
        else:
            pending_section = (
                f"### Pendiente de clasificación\n"
                f"- {pending_count} sesiones pendientes de clasificación"
            )
        if body:
            body = f"{body}\n\n{pending_section}"
        else:
            body = pending_section

    return body


def _template_slack(
    activities: list[Activity],
    pending_count: int,
    lang: str,
) -> str:
    """Render activities as a Slack-formatted template report."""
    sections: list[str] = []

    sorted_acts = sorted(activities, key=operator.attrgetter("day"))
    for day, group in itertools.groupby(sorted_acts, key=operator.attrgetter("day")):
        lines: list[str] = [f"*{day}*", ""]
        lines.append("*Done*" if lang == "en" else "*Done*")
        for act in group:
            lines.append(f"\u2022 {_format_activity_line(act)}")
        sections.append("\n".join(lines))

    body = "\n\n".join(sections)

    if pending_count > 0:
        if lang == "en":
            pending_section = (
                f"*Pending classification*\n"
                f"\u2022 {pending_count} sessions pending classification"
            )
        else:
            pending_section = (
                f"*Pendiente de clasificación*\n"
                f"\u2022 {pending_count} sesiones pendientes de clasificación"
            )
        if body:
            body = f"{body}\n\n{pending_section}"
        else:
            body = pending_section

    return body


def generate_template_report(
    activities: list[Activity],
    output_format: str = "markdown",
    pending_count: int = 0,
    lang: str = "es",
) -> str:
    """Generate an instant standup report — no LLM calls, pure template."""
    if not activities and pending_count <= 0:
        if lang == "en":
            return "No activity found for the requested period."
        return "No se encontró actividad para el período solicitado."

    if output_format == "slack":
        return _template_slack(activities, pending_count, lang)
    return _template_markdown(activities, pending_count, lang)


# ---------------------------------------------------------------------------
# LLM-polished report — one `claude -p` call
# ---------------------------------------------------------------------------

_LLM_REPORT_PROMPT = """\
You are generating a daily standup report for a software developer.

Below is raw classified activity data from their Claude Code sessions. Your job:

1. Write a clean, concise activity summary. One section: what was done, grouped by project.
2. IGNORE noise: subagent instructions ("You are implementing...", "You are reviewing..."), \
   internal tool output ("List project directory", "Check git status"), JSON artifacts, and \
   anything that looks like Claude Code internal machinery rather than actual developer work
3. MERGE only truly redundant entries (exact same task mentioned multiple times). \
   When there are multiple distinct deliverables (e.g. several PRs, several services, \
   several bugs fixed), LIST EACH ONE by name. Never collapse distinct items into \
   a single vague summary. Bad: "Created new inference services". \
   Good: "Created kling-4k-t2v and kling-4k-i2v inference services". \
   Extract specific names from the context data (tool descriptions, assistant messages).
4. Include org/repo when available. Do NOT include time estimates or durations.
5. Write in {lang}
6. Use {format} formatting
7. Do NOT include "Next steps", "Blockers", or "TODO" sections. Only summarize what was done.
{context_section}
Raw activity data:
{activities}
"""


def generate_llm_report(
    activities: list[Activity],
    output_format: str = "markdown",
    lang: str = "es",
    context_activities: list[Activity] | None = None,
) -> str:
    """Generate a polished standup report using one claude -p call.

    *context_activities*: activities from previous days (e.g. yesterday) so the LLM
    knows what's already done and doesn't suggest completed work as "Next".
    """
    if not activities:
        if lang == "en":
            return "No activity found for the requested period."
        return "No se encontró actividad para el período solicitado."

    # Build raw data for the LLM — include all prompts so it can see completion signals
    lines = []
    for act in activities:
        org_repo = f"({act.git_org}/{act.git_repo})" if act.git_org and act.git_repo else ""
        lines.append(f"- [{act.day}] [{act.classification}]{org_repo} {act.summary}")
        # Include raw prompts (truncated) so LLM can see the full story
        for prompt in act.raw_prompts[:15]:  # max 15 context lines per activity
            truncated = prompt.strip().split("\n")[0][:120]
            if len(truncated) > 15:
                lines.append(f"    > {truncated}")

    # Build context section from previous days
    context_section = ""
    if context_activities:
        ctx_lines = []
        for act in context_activities:
            org_repo = f"({act.git_org}/{act.git_repo})" if act.git_org and act.git_repo else ""
            ctx_lines.append(f"- [{act.day}] [{act.classification}]{org_repo} {act.summary}")
        context_section = (
            "\nContext — activities from previous days (already completed, do NOT suggest as Next):\n"
            + "\n".join(ctx_lines) + "\n\n"
        )

    lang_name = "Spanish" if lang == "es" else "English"
    fmt_name = "Slack (*bold*, • bullets)" if output_format == "slack" else "Markdown (## headers, - bullets)"

    prompt = _LLM_REPORT_PROMPT.format(
        lang=lang_name,
        format=fmt_name,
        activities="\n".join(lines),
        context_section=context_section,
    )

    claude_path = shutil.which("claude")
    if not claude_path:
        # Fallback to template if claude not available
        return generate_template_report(activities, output_format=output_format, lang=lang)

    try:
        result = subprocess.run(
            [claude_path, "-p", prompt, "--output-format", "text", "--no-session-persistence"],
            capture_output=True, text=True, timeout=60,
        )
        if result.returncode == 0 and result.stdout.strip():
            return result.stdout.strip()
    except (subprocess.TimeoutExpired, OSError):
        pass

    # Fallback to template on any failure
    return generate_template_report(activities, output_format=output_format, lang=lang)

File: src/reporter.rs
use std::collections::BTreeMap;
use std::process::Command;

use crate::models::Activity;

/// Format a single activity line.
///
/// Pattern: `[classification](org/repo) summary`
/// - Skip classification prefix if empty.
/// - Skip org/repo if both are None.
fn format_activity_line(activity: &Activity, bullet: &str) -> String {
    let mut parts = Vec::new();

    if !activity.classification.is_empty() {
        parts.push(format!("[{}]", activity.classification));
    }

    let has_org = activity.git_org.as_ref().is_some_and(|s| !s.is_empty());
    let has_repo = activity.git_repo.as_ref().is_some_and(|s| !s.is_empty());

    if has_org || has_repo {
        let org = activity.git_org.as_deref().unwrap_or("");
        let repo = activity.git_repo.as_deref().unwrap_or("");
        parts.push(format!("({}/{})", org, repo));
    }

    parts.push(activity.summary.clone());

    format!("{} {}", bullet, parts.join(" "))
}

/// Generate a template-based standup report (no LLM).
///
/// Groups activities by day (sorted ascending). Renders in Markdown or Slack format.
pub fn generate_template_report(activities: &[Activity], format: &str, lang: &str) -> String {
    if activities.is_empty() {
        return match lang {
            "es" => "No se encontró actividad para el período solicitado.".to_string(),
            _ => "No activity found for the requested period.".to_string(),
        };
    }

    // Group activities by day using a BTreeMap for sorted keys.
    let mut by_day: BTreeMap<&str, Vec<&Activity>> = BTreeMap::new();
    for activity in activities {
        by_day.entry(activity.day.as_str()).or_default().push(activity);
    }

    let is_slack = format == "slack";
    let mut output = String::new();

    for (day, day_activities) in &by_day {
        // Day header
        if is_slack {
            output.push_str(&format!("*{}*\n\n", day));
        } else {
            output.push_str(&format!("## {}\n\n", day));
        }

        // "Done" section header
        if is_slack {
            output.push_str("*Done*\n");
        } else {
            output.push_str("### Done\n");
        }

        // Activity lines
        let bullet = if is_slack { "•" } else { "-" };
        for activity in day_activities {
            output.push_str(&format_activity_line(activity, bullet));
            output.push('\n');
        }

        output.push('\n');
    }

    output.trim_end().to_string()
}

/// Generate an LLM-powered standup report by calling `claude -p`.
///
/// Falls back to `generate_template_report` if `claude` is not available or fails.
pub fn generate_llm_report(activities: &[Activity], format: &str, lang: &str) -> String {
    if activities.is_empty() {
        return generate_template_report(activities, format, lang);
    }

    // Build raw activity text lines.
    let activities_text: String = activities
        .iter()
        .map(|a| {
            let has_org = a.git_org.as_ref().is_some_and(|s| !s.is_empty());
            let has_repo = a.git_repo.as_ref().is_some_and(|s| !s.is_empty());

            let repo_part = if has_org || has_repo {
                let org = a.git_org.as_deref().unwrap_or("");
                let repo = a.git_repo.as_deref().unwrap_or("");
                format!("({}/{})", org, repo)
            } else {
                String::new()
            };

            let class_part = if a.classification.is_empty() {
                String::new()
            } else {
                format!("[{}]", a.classification)
            };

            format!(
                "- [{}] {}{} {}",
                a.day,
                class_part,
                repo_part,
                a.summary
            )
        })
        .collect::<Vec<_>>()
        .join("\n");

    let lang_name = match lang {
        "es" => "Spanish",
        _ => "English",
    };

    let format_desc = if format == "slack" {
        "Slack (*bold*, • bullets)"
    } else {
        "Markdown (## headers, - bullets)"
    };

    let prompt = format!(
        r#"You are generating a daily standup report for a software developer.

Below is raw classified activity data from their Claude Code sessions. Your job:

1. Write a clean, concise activity summary. Group by project. Only what was done.
2. IGNORE noise and mechanical/routine tasks that add no value to a standup: subagent instructions, internal tool output, git sync/rebase/pull, branch cleanup, submodule updates, pushing pointers, adding reviewers, linting, formatting.
3. When there are multiple distinct deliverables (PRs, services, bugs), LIST EACH by name.
4. Keep bullets concise. One line per deliverable. No implementation details.
5. Include org/repo when available. No time estimates.
6. Multiple related fixes to the same component = one bullet summarizing the outcome.
7. Write in {lang}. Use {format} formatting.
8. Do NOT include "Next steps", "Blockers", or "TODO" sections.

Raw activity data:
{activities}"#,
        lang = lang_name,
        format = format_desc,
        activities = activities_text,
    );

    let result = Command::new("claude")
        .args(["-p", &prompt, "--output-format", "text", "--no-session-persistence"])
        .output();

    match result {
        Ok(output) if output.status.success() => {
            let text = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if text.is_empty() {
                generate_template_report(activities, format, lang)
            } else {
                text
            }
        }
        _ => generate_template_report(activities, format, lang),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::Activity;

    fn sample_activities() -> Vec<Activity> {
        vec![
            Activity {
                session_id: "s1".to_string(),
                day: "2026-04-20".to_string(),
                project: "api".to_string(),
                git_org: Some("acme".to_string()),
                git_repo: Some("api".to_string()),
                classification: "feature".to_string(),
                summary: "Add user endpoint".to_string(),
            },
            Activity {
                session_id: "s2".to_string(),
                day: "2026-04-20".to_string(),
                project: "api".to_string(),
                git_org: Some("acme".to_string()),
                git_repo: Some("api".to_string()),
                classification: "bugfix".to_string(),
                summary: "Fix auth middleware".to_string(),
            },
            Activity {
                session_id: "s3".to_string(),
                day: "2026-04-21".to_string(),
                project: "web".to_string(),
                git_org: None,
                git_repo: None,
                classification: "".to_string(),
                summary: "Update landing page".to_string(),
            },
        ]
    }

    #[test]
    fn test_empty_activities_en() {
        let result = generate_template_report(&[], "markdown", "en");
        assert_eq!(result, "No activity found for the requested period.");
    }

    #[test]
    fn test_empty_activities_es() {
        let result = generate_template_report(&[], "markdown", "es");
        assert_eq!(
            result,
            "No se encontró actividad para el período solicitado."
        );
    }

    #[test]
    fn test_markdown_format() {
        let activities = sample_activities();
        let result = generate_template_report(&activities, "markdown", "en");

        assert!(result.contains("## 2026-04-20"));
        assert!(result.contains("## 2026-04-21"));
        assert!(result.contains("### Done"));
        assert!(result.contains("- [feature] (acme/api) Add user endpoint"));
        assert!(result.contains("- [bugfix] (acme/api) Fix auth middleware"));
        assert!(result.contains("- Update landing page"));
    }

    #[test]
    fn test_slack_format() {
        let activities = sample_activities();
        let result = generate_template_report(&activities, "slack", "en");

        assert!(result.contains("*2026-04-20*"));
        assert!(result.contains("*2026-04-21*"));
        assert!(result.contains("*Done*"));
        assert!(result.contains("• [feature] (acme/api) Add user endpoint"));
        assert!(result.contains("• Update landing page"));
    }

    #[test]
    fn test_days_sorted() {
        let activities = vec![
            Activity {
                session_id: "s1".to_string(),
                day: "2026-04-21".to_string(),
                project: "b".to_string(),
                git_org: None,
                git_repo: None,
                classification: "".to_string(),
                summary: "Second day".to_string(),
            },
            Activity {
                session_id: "s2".to_string(),
                day: "2026-04-19".to_string(),
                project: "a".to_string(),
                git_org: None,
                git_repo: None,
                classification: "".to_string(),
                summary: "First day".to_string(),
            },
        ];
        let result = generate_template_report(&activities, "markdown", "en");
        let pos_19 = result.find("2026-04-19").unwrap();
        let pos_21 = result.find("2026-04-21").unwrap();
        assert!(pos_19 < pos_21, "Days should be sorted ascending");
    }

    #[test]
    fn test_llm_report_empty_fallback() {
        let result = generate_llm_report(&[], "markdown", "en");
        assert_eq!(result, "No activity found for the requested period.");
    }
}
